dbms/src/DataStreams/ParallelParsingBlockInputStream.cpp

```diff
@@ -0,0 +1,145 @@
+#include <DataStreams/ParallelParsingBlockInputStream.h>
+
+namespace DB
+{
+
+void ParallelParsingBlockInputStream::segmentatorThreadFunction()
+{
+    setThreadName("Segmentator");
+    try
+    {
+        while (!is_cancelled && !is_exception_occured)
+        {
+            ++segmentator_ticket_number;
+            const auto current_unit_number = segmentator_ticket_number % max_threads_to_use;
+            auto & current_unit = working_field[current_unit_number];
+
+            {
+                std::unique_lock lock(mutex);
+                segmentator_condvar.wait(lock, [&](){ return current_unit.status == READY_TO_INSERT || is_exception_occured || is_cancelled; });
+            }
+
+            if (is_exception_occured)
+                break;
+
+            if (original_buffer.eof())
+            {
+                std::unique_lock lock(mutex);
+                current_unit.is_last_unit = true;
+                current_unit.status = READY_TO_PARSE;
+                scheduleParserThreadForUnitWithNumber(current_unit_number);
+                break;
+            }
+
+            // Segmentating the original input.
+            current_unit.chunk.used_size = 0;
+            bool res = file_segmentation_engine(original_buffer, current_unit.chunk.memory, current_unit.chunk.used_size, min_chunk_size);
+
+            // Creating buffer from the segment of data.
+            auto new_buffer = BufferBase::Buffer(current_unit.chunk.memory.data(), current_unit.chunk.memory.data() + current_unit.chunk.used_size);
+            current_unit.readbuffer.buffer().swap(new_buffer);
+            current_unit.readbuffer.position() = current_unit.readbuffer.buffer().begin();
+
+            if (!res)
+            {
+                std::unique_lock lock(mutex);
+                current_unit.is_last_unit = true;
+                current_unit.status = READY_TO_PARSE;
+                scheduleParserThreadForUnitWithNumber(current_unit_number);
+                break;
+            }
+
+            std::unique_lock lock(mutex);
+            current_unit.status = READY_TO_PARSE;
+            scheduleParserThreadForUnitWithNumber(current_unit_number);
```

Comment:
Looks like we don't need the lock here -- schedule() call is a sufficient barrier to make the current_unit changes seen by the parser thread. A caveat: to conform to the C++ standard, current_unit.status should be std::atomic<>. A concurrent read may also occur from reader thread, and without atomic<> it would be UB.